src/client_builder.rs

```diff
@@ -116,19 +237,23 @@ where
     /// # fn main() -> Result<(), imap::Error> {
     /// let client = ClientBuilder::new("imap.example.com", 993)
     ///     .starttls()
-    ///     .connect(|domain, tcp| {
+    ///     .connect_tls_with(|domain, tcp| {
     ///         let ssl_conn = RustlsConnector::new_with_native_certs()?;
     ///         Ok(ssl_conn.connect(domain, tcp)?)
     ///     })?;
     /// # Ok(())
     /// # }
     /// ```
-    pub fn connect<F, C>(&mut self, handshake: F) -> Result<Client<C>>
+    pub fn connect_tls_with<F, C>(&mut self, handshake: F) -> Result<Client<C>>
     where
         F: FnOnce(&str, TcpStream) -> Result<C>,
         C: Read + Write,
     {
-        let tcp = if self.starttls {
+        if self.mode == ConnectionMode::Tcp {
+            panic!("Can only be called with a TLS or StartTLS connection");
+        }
```

Comment:
This doesn't seem right to me. My expectation would actually be that if someone calls `connect_with` and hasn't called `tls()`, then this should work just fine and just use exactly what comes out of the `FnOnce` as the channel. This suggests, perhaps, that `Tcp` isn't the right name for this mode, and that it should instead be called something like `Direct` or `Plain` or `Unmodified`.

Reply:
Per your comment about the function name, I'm just removing this check completely.